src/equations/ideal_mhd_multiion_2d.jl

```diff
@@ -0,0 +1,916 @@
+    
+    v1 = rho_v1 / rho
+    v2 = rho_v2 / rho
+    v3 = rho_v3 / rho
+    v_mag = sqrt(v1^2 + v2^2 + v3^2)
+    gamma = equations.gammas[k]
+    p = (gamma - 1)*(rho_e - 0.5*rho*v_mag^2 - 0.5*(B1^2 + B2^2 + B3^2))
+    a_square = gamma * p / rho
+    sqrt_rho = sqrt(rho)
+
+    b1 = B1 / sqrt_rho
+    b2 = B2 / sqrt_rho
+    b3 = B3 / sqrt_rho
+    b_square = b1^2 + b2^2 + b3^2
+
+    if orientation == 1
+      c_f = max(c_f, sqrt(0.5*(a_square + b_square) + 0.5*sqrt((a_square + b_square)^2 - 4.0*a_square*b1^2)))
+    else #if orientation == 2
+      c_f = max(c_f, sqrt(0.5*(a_square + b_square) + 0.5*sqrt((a_square + b_square)^2 - 4.0*a_square*b2^2)))
+    end
+  end
+
+  return c_f
+end
+
+"""
+Routine to compute the Charge-averaged velocities:
+* v*_plus: Charge-averaged velocity
+* vk*_plus: Contribution of each species to the charge-averaged velocity
+"""
+@inline function charge_averaged_velocities(u, equations::IdealMhdMultiIonEquations2D)
+
+  total_electron_charge = zero(real(equations))
+  
+  vk1_plus = zero(MVector{ncomponents(equations), eltype(u)})
+  vk2_plus = zero(MVector{ncomponents(equations), eltype(u)})
+  vk3_plus = zero(MVector{ncomponents(equations), eltype(u)})
+
+  for k in eachcomponent(equations)
+    rho_k = u[(k-1)*5+4]
+    rho_v1_k = u[(k-1)*5+5]
+    rho_v2_k = u[(k-1)*5+6]
+    rho_v3_k = u[(k-1)*5+7]
+    total_electron_charge += rho_k * equations.charge_to_mass[k]
+    vk1_plus[k] = rho_v1_k * equations.charge_to_mass[k]
+    vk2_plus[k] = rho_v2_k * equations.charge_to_mass[k]
+    vk3_plus[k] = rho_v3_k * equations.charge_to_mass[k]
+  end
+  vk1_plus ./= total_electron_charge
+  vk2_plus ./= total_electron_charge
+  vk3_plus ./= total_electron_charge
+  v1_plus = sum(vk1_plus)
+  v2_plus = sum(vk2_plus)
+  v3_plus = sum(vk3_plus)
+
+  return v1_plus, v2_plus, v3_plus, SVector(vk1_plus), SVector(vk2_plus), SVector(vk3_plus)
+end
+
+"""
+Get the flow variables of component k
+"""
+@inline function get_component(k, u, equations::IdealMhdMultiIonEquations2D)
+  # The first 3 entries of u contain the magnetic field. The following entries contain the density, momentum (3 entries), and energy of each component.
+  return SVector(u[3 + (k - 1) * 5 + 1],
+                 u[3 + (k - 1) * 5 + 2],
+                 u[3 + (k - 1) * 5 + 3],
+                 u[3 + (k - 1) * 5 + 4],
+                 u[3 + (k - 1) * 5 + 5])
+end
+
+"""
+Set the flow variables of component k
+"""
+@inline function set_component!(u, k, u1, u2, u3, u4, u5, equations::IdealMhdMultiIonEquations2D)
+  # The first 3 entries of u contain the magnetic field. The following entries contain the density, momentum (3 entries), and energy of each component.
+  u[3 + (k - 1) * 5 + 1] = u1
+  u[3 + (k - 1) * 5 + 2] = u2
+  u[3 + (k - 1) * 5 + 3] = u3
+  u[3 + (k - 1) * 5 + 4] = u4
+  u[3 + (k - 1) * 5 + 5] = u5
+end
+
+@inline function density_product(u, equations::IdealMhdMultiIonEquations2D)
+  prod = one(u[1])
```

Comment:
```suggestion
  prod = one(eltype(u))
```
I find this clearer since it does not imply that `u[1]` is somehow special (applies also elsewhere in this file)